django/db/migrations/loader.py

```diff
@@ -174,92 +198,55 @@ def build_graph(self):
         else:
             recorder = MigrationRecorder(self.connection)
             self.applied_migrations = recorder.applied_migrations()
-        # Do a first pass to separate out replacing and non-replacing migrations
-        normal = {}
-        replacing = {}
+        # To start, populate the migration graph with nodes for ALL migrations
+        # and their dependencies. Also make note of replacing migrations at this step.
+        self.graph = MigrationGraph()
+        self.replacements = {}
         for key, migration in self.disk_migrations.items():
+            self.graph.add_node(key, migration)
+            # Internal (aka same-app) dependencies.
+            self.add_internal_dependencies(key, migration)
+            # Replacing migrations.
             if migration.replaces:
-                replacing[key] = migration
-            else:
-                normal[key] = migration
-        # Calculate reverse dependencies - i.e., for each migration, what depends on it?
-        # This is just for dependency re-pointing when applying replacements,
-        # so we ignore run_before here.
-        reverse_dependencies = {}
-        for key, migration in normal.items():
-            for parent in migration.dependencies:
-                reverse_dependencies.setdefault(parent, set()).add(key)
-        # Remember the possible replacements to generate more meaningful error
-        # messages
-        reverse_replacements = {}
-        for key, migration in replacing.items():
-            for replaced in migration.replaces:
-                reverse_replacements.setdefault(replaced, set()).add(key)
-        # Carry out replacements if we can - that is, if all replaced migrations
-        # are either unapplied or missing.
-        for key, migration in replacing.items():
-            # Ensure this replacement migration is not in applied_migrations
-            self.applied_migrations.discard(key)
-            # Do the check. We can replace if all our replace targets are
-            # applied, or if all of them are unapplied.
+                self.replacements[key] = migration
+        # Add external dependencies now that the internal ones have been resolved.
+        for key, migration in self.disk_migrations.items():
+            self.add_external_dependencies(key, migration)
+        # Carry out replacements where possible.
```

Comment:
There could be deleted replacement targets with a replacing migration still has them in its `replaces` list. In this case other normal migrations are still allowed to depend on these deleted migrations, but if they do, they'll be creating `DummyNode`s that won't be resolved until after the next section runs.